unity/Assets/Scripts/PhysicsRemoteFPSAgentController.cs

```diff
@@ -4476,35 +4476,55 @@ protected void SetStateOfObjectsThatDontHaveAnimationTime(SetObjectStates SetObj
             actionFinished(true);
         }
 
-        public void PutObject(float x, float y, bool forceAction=false, bool placeStationary=true){
-            PlaceHeldObject(x, y, forceAction, placeStationary);
+        public void PutObject(float x, float y, bool forceAction=false, bool placeStationary=true, int randomSeed=0, bool putNearXY=false){
+            PlaceHeldObject(
+                x: x, 
+                y: y,
+                forceAction: forceAction, 
+                placeStationary: placeStationary,
+                randomSeed: randomSeed,
+                putNearXY: putNearXY);
         }
 
-        public void PutObject(string objectId, bool forceAction=false, bool placeStationary=true){
-            PlaceHeldObject(objectId, forceAction, placeStationary);
+        public void PutObject(string objectId, bool forceAction=false, bool placeStationary=true, int randomSeed=0){
+            PlaceHeldObject(                
+                objectId: objectId,
+                forceAction: forceAction,
+                placeStationary: placeStationary,
+                randomSeed: randomSeed);
         }
 
         //if you are holding an object, place it on a valid Receptacle 
         //used for placing objects on receptacles without enclosed restrictions (drawers, cabinets, etc)
         //only checks if the object can be placed on top of the target receptacle
-        public void PlaceHeldObject(float x, float y, bool forceAction=false, bool placeStationary=true, int randomSeed = 0, float z = 0.0f){
+        public void PlaceHeldObject(float x, float y, bool forceAction=false, bool placeStationary=true, int randomSeed = 0, float maxDistance = 0.0f, bool putNearXY = false){
```

Comment:
As changing `z` to `maxDistance` is an backwards incompatible API change we should only be making this at the next larger release (e.g. 2.8).

Reply:
This shouldn't be an issue for backwards compatibility. The `maxDistance`/`z` value is only internally used by the `PlaceHeldObject` function. `PlaceHeldObject` is onlyu called by `PutObject` which never exposed the  `maxDistance`/`z` to the python controller.

The original versions of `PutObject` were:

```       
        //Target what receptacle to put a held object via (x, y) screen coordinates
        public void PutObject(float x, float y, bool forceAction=false, bool placeStationary=true){
            PlaceHeldObject(x, y, forceAction, placeStationary);
        }

        //Target what receptacle to put a held object via objectId directly
        public void PutObject(string objectId, bool forceAction=false, bool placeStationary=true){
            PlaceHeldObject(objectId, forceAction, placeStationary);
        }
 ```
 
The new versions included in this update still don't expose or change the default values of `maxDistance`/`z` , but they do add new optional parameters of `randomSeed` and `putNearXY`

```        
        public void PutObject(float x, float y, bool forceAction=false, bool placeStationary=true, int 
        randomSeed=0, bool putNearXY=false){
            PlaceHeldObject(
                x: x, 
                y: y,
                forceAction: forceAction, 
                placeStationary: placeStationary,
                randomSeed: randomSeed,
                putNearXY: putNearXY);
        }

        public void PutObject(string objectId, bool forceAction=false, bool placeStationary=true, 
        int randomSeed=0){
            PlaceHeldObject(                
                objectId: objectId,
                forceAction: forceAction,
                placeStationary: placeStationary,
                randomSeed: randomSeed);
        }
```

Reply:
I'm not sure I understand, `PlaceHeldObject` is exposed to the Python API and so anyone who used this action with the `z` argument would now find that this doesn't work. Or am I wrong?

Reply:
Yes, it is exposed but the documented one that should be used directly from the Python API is [PutObject](https://ai2thor.allenai.org/ithor/documentation/interactive-physics/#sub-put-object)

`PutObject` has not changed and never took `z`. Although I'm not sure if there are older projects that instead use `PlaceHeldObject` directly, which may be the concern here.

Reply:
Yes that's my concern. Generally I'm ok with such changes so long as they throw an error, can you create a python-side test that checks that taking a `PlaceHeldObject` breaks when called with `z` but succeeds when called with `maxDistance` (all other parameters the same)?

Reply:
yes I'll try this. I think because of the new action dispatcher, passing in a param that is not used will throw some sort of exception? I'll update here once I try it out.

Reply:
ok! so after some back and forth, if someone happens to be using `PlaceHeldObject` and passes in `z`, a missing argument exception now gets thrown:

```Action failed with error message 'action: PlaceHeldObject is missing the following arguments: maxDistance```

There is still a potential issue where if one were to pass in both `maxDistance` and `z` it would succeed with potentially silent unintended behavior, but we need to have a larger conversation about accounting for this later. As both Eric and Alvaro brought up, we currently don't want to have a forced exception check for extraneous parameters.

To facilitate this, I've made `maxDistance` a required parameter on `PlaceHeldObject` and the defaulting of `maxDistance` to equal `maxVisibleDistance` set on initialize is now done within `PutObject`. So `PutObject` behaves exactly the same as before even though `maxDistance` is no longer a nullable type.

Additionally added in a bunch of unit tests to help check for all combinations of this.